File: bug71_incorrect_reverse_loop/bug71_incorrect_reverse_loop/main.java
public class main {
    public static void main(String[] args) {
        long[] data = new long[10];
        for (int i = 0; i < 10; i++) data[i] = 0x100L * (i + 1);
        
        long acc = 0;
        for (int i = 9; i > 0; i--) {
            acc ^= data[i] * (i + 1);
            acc = (acc * 0x9E3779B97F4A7C15L);
            acc = ((acc << (i % 7 + 1)) | (acc >>> (63 - (i % 7))));
        }
        
        long correct = 0;
        for (int i = 9; i >= 0; i--) {
            correct ^= data[i] * (i + 1);
            correct = (correct * 0x9E3779B97F4A7C15L);
            correct = ((correct << (i % 7 + 1)) | (correct >>> (63 - (i % 7))));
        }
        
        if (acc == correct) {
            System.err.println("AssertionError: reverse loop stop same result");
            System.exit(1);
        }
        
        System.err.printf("ValueError: reverse loop stops at 1 not 0, skips index 0: got 0x%s expected 0x%s\n", Long.toHexString(acc), Long.toHexString(correct));
        System.exit(1);
    }
}

File: bug71_incorrect_reverse_loop/bug71_incorrect_reverse_loop/main.js
function main() {
    let data = [];
    for (let i = 0n; i < 10n; i++) data.push(0x100n * (i + 1n));
    
    let acc = 0n;
    for (let j = 9; j > 0; j--) {
        let i = BigInt(j);
        acc ^= data[j] * (i + 1n);
        acc = (acc * 0x9E3779B97F4A7C15n) & 0xFFFFFFFFFFFFFFFFn;
        acc = ((acc << (i % 7n + 1n)) | (acc >> (63n - i % 7n))) & 0xFFFFFFFFFFFFFFFFn;
    }
    
    let correct = 0n;
    for (let j = 9; j >= 0; j--) {
        let i = BigInt(j);
        correct ^= data[j] * (i + 1n);
        correct = (correct * 0x9E3779B97F4A7C15n) & 0xFFFFFFFFFFFFFFFFn;
        correct = ((correct << (i % 7n + 1n)) | (correct >> (63n - i % 7n))) & 0xFFFFFFFFFFFFFFFFn;
    }
    
    if (acc === correct) {
        console.error("AssertionError: reverse loop stop same result");
        process.exit(1);
    }
    console.error(`ValueError: reverse loop stops at 1 not 0, skips index 0: got 0x${acc.toString(16)} expected 0x${correct.toString(16)}`);
    process.exit(1);
}
main();
